### ISSUE TITLE
TypeScript Definitions: fix Knex.QueryBuilder.extend signature

### ISSUE THREAD

**Issue by iki** [knex/knex#3526]
# Environment

Knex version: 0.19.5

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

Function passed to `Knex.QueryBuilder.extend` uses `this` as `QueryBuilder` . However it is typed as `Knex`.

2. Error message

TypeScript complains about `QueryBuilder` methods like `toSQL` or Promise methods like `then` when used on `this` in the extending function.

```Property 'then' does not exist on type Knex<any, any[]>```

3. Reduced test code

See https://github.com/knex/knex/issues/1355#issuecomment-551435034.

It uses `this: Knex.QueryBuilder` in function signature to fix the problem and types the fuction as `any` in the end to avoid TypeScript complaining about unassignable function types due to `Knex` and `QueryBuilder` type incompatibility.

4. Solution

The [`Knex.QueryBuilder.extend` signature](https://github.com/knex/knex/blob/master/types/index.d.ts#L1930-L1936) should use `QueryBuilder` for `this`:
```ts
static extend(
    methodName: string,
    fn: <TRecord extends {} = any, TResult = unknown[]>(
      this: QueryBuilder<TRecord, TResult>,
      ...args: any[]
    ) => QueryBuilder<TRecord, TResult>
  ): void;
```

instead of `Knex`:
```ts
static extend(
    methodName: string,
    fn: <TRecord extends {} = any, TResult = unknown[]>(
      this: Knex<TRecord, TResult>,
      ...args: any[]
    ) => QueryBuilder<TRecord, TResult>
  ): void;
```

**Comment by lorefnon** [comment]
@iki Since you have already figured out the solution, would you consider creating a PR for this ?

**Comment by kibertoad** [comment]
Released in 0.20.4
